src/lib/gc.js

```diff
@@ -0,0 +1,229 @@
+ * can also register callback functions to do arbitrary cleanup upon teardown.
+ *
+ * Henrik Ingo (c) 2016
+ * MIT License
+ */
+
+( function( document, window ) {
+    "use strict";
+    var roots = [];
+    var rootsCount = 0;
+    var startingState = { roots: [] };
+
+    var libraryFactory = function( rootId ) {
+        if ( roots[ rootId ] ) {
+            return roots[ rootId ];
+        }
+
+        // Per root global variables (instance variables?)
+        var elementList = [];
+        var eventListenerList = [];
+        var callbackList = [];
+
+        recordStartingState( rootId );
+
+        // LIBRARY FUNCTIONS
+        // Below are definitions of the library functions we return at the end
+        var pushElement = function( element ) {
+            elementList.push( element );
+        };
+
+        // Convenience wrapper that combines DOM appendChild with gc.pushElement
+        var appendChild = function( parent, element ) {
+            parent.appendChild( element );
+            pushElement( element );
+        };
+
+        var pushEventListener = function( target, type, listenerFunction ) {
+            eventListenerList.push( { target:target, type:type, listener:listenerFunction } );
+        };
+
+        // Convenience wrapper that combines DOM addEventListener with gc.pushEventListener
+        var addEventListener = function( target, type, listenerFunction ) {
+            target.addEventListener( type, listenerFunction );
+            pushEventListener( target, type, listenerFunction );
+        };
+
+        // If the above utilities are not enough, plugins can add their own callback function
+        // to do arbitrary things.
+        var addCallback = function( callback ) {
+            callbackList.push( callback );
+        };
+        addCallback( function( rootId ) { resetStartingState( rootId ); } );
+
+        var teardown = function() {
+
+            // Execute the callbacks in LIFO order
+            var i; // Needed by jshint
+            for ( i = callbackList.length - 1; i >= 0; i-- ) {
+                callbackList[ i ]( rootId );
+            }
+            callbackList = [];
+            for ( i = 0; i < elementList.length; i++ ) {
+                elementList[ i ].parentElement.removeChild( elementList[ i ] );
+            }
+            elementList = [];
+            for ( i = 0; i < eventListenerList.length; i++ ) {
+                var target   = eventListenerList[ i ].target;
+                var type     = eventListenerList[ i ].type;
+                var listener = eventListenerList[ i ].listener;
+                target.removeEventListener( type, listener );
+            }
+        };
+
+        var lib = {
+            pushElement: pushElement,
+            appendChild: appendChild,
+            pushEventListener: pushEventListener,
+            addEventListener: addEventListener,
+            addCallback: addCallback,
```

Comment:
@henrikingo I think my biggest complain or worry in this PR will be this GC API.

First of all (already mentioned elsewhere) the name 'garbage collector' is a bit confusing and it would be good to rename it at some point (before we expose it publicly).

For the methods it has in API - it's really hard to tell the difference between `pushEventListener` and `addEventListener` without looking at the code of them.
I'm not sure how flexible we need this API.

In theory, the only 'required' thing is `addCallback` - so that plugins can add a callback and clean up after themselves on teardown. Any other method (registering elements, events, etc) are for convenience.
I don't know how 'convenient' we need this API to be as it already seems confusing.
But maybe what's confusing is names of the functions that are very similar to each other and it's hard to say what they do differently.

I don't know if it's good that we mix low level 'registry' functions like `pushElement`, `pushEventListener` `addCallback` with wrappers over DOM methods `appendChild` `addEventListener`.

If we assume that plugins should use convenience methods for DOM operations, maybe we should expose only those.
If we want plugins to do their own DOM operations and only register elements/events, maybe low level functions would be enough?

I'm trying to put myself in a position of potential plugin developer - and looking at this API I don't know what I'm meant to be using.
I kind of feel it's about the balance of being more friendly to less experienced developer (having simpler API) or to more experienced one (having both convenience and low level functions).

I don't have a good answer or suggestion, I just feel that in current form it may be confusing.

@henrikingo What are your thoughts on that?

Reply:
> I'm trying to put myself in a position of potential plugin developer - and looking at this API I don't know what I'm meant to be using.

From my experience writing 20 plugins for impress.js, and also using same mechanism in impressionist:

* I want methods that wrap around the DOM methods I used until gc was introduced.
* Sometimes I use some other JavaScript libraries that create DOM elements, and I just need to add them to gc so they are removed on teardown. This is what the push* functions do.
* addCallback() is the catch all where a plugin can do anything. But, expecting each plugin to do the addEventListener and appendChild work separately, would be just as annoying and bad style as copy pasting lib.util.triggerEvent to each plugin.
* There's also a need for add* and push* method that would just set and reset an attribute value. This doesn't exist yet.

Reply:
So what's the final decision, leave the GC as it is and take a look later?